orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -52,66 +63,76 @@ public ArtifactResolver(ObjectMapper objectMapper, ExecutionRepository execution
     this.executionRepository = executionRepository;
   }
 
-  public @Nonnull List<Artifact> getArtifacts(@Nonnull Stage stage) {
+  public @Nonnull
+  List<Artifact> getArtifacts(@Nonnull Stage stage) {
     if (stage.getContext() instanceof StageContext) {
-      return ((StageContext) stage.getContext()).getAll("artifacts");
+      return (List<Artifact>) Optional.ofNullable((List) ((StageContext) stage.getContext()).getAll("artifacts"))
+          .map(list -> list.stream()
+            .map(it -> objectMapper.convertValue(it, new TypeReference<List<Artifact>>() {}))
+            .filter(Objects::nonNull)
+            .flatMap(it -> ((List) it).stream())
+            .collect(Collectors.toList()))
+          .orElse(emptyList());
     } else {
       log.warn("Unable to read artifacts from unknown context type: {} ({})", stage.getContext().getClass(), stage.getExecution().getId());
       return emptyList();
     }
   }
 
-  public @Nullable Artifact getBoundArtifactForId(
-    @Nonnull Stage stage, @Nullable String id) {
+  public @Nullable
+  Artifact getBoundArtifactForId(
+      @Nonnull Stage stage, @Nullable String id) {
     if (isEmpty(id)) {
       return null;
     }
 
     List<ExpectedArtifact> expectedArtifacts;
     if (stage.getContext() instanceof StageContext) {
-      expectedArtifacts = ((StageContext) stage.getContext()).getAll("resolvedExpectedArtifacts");
+      expectedArtifacts = (List<ExpectedArtifact>) Optional.ofNullable((List) ((StageContext) stage.getContext()).getAll("resolvedExpectedArtifacts"))
+          .map(list -> list.stream()
+              .map(it -> objectMapper.convertValue(it, new TypeReference<List<ExpectedArtifact>>() {}))
```

Comment:
Feels like we should be able to avoid this by declaring the property on the trigger correctly. Will take a proper look when I get to work.